src/System.Net.Security/src/System/Net/_SSPIWrapper.cs

```diff
@@ -0,0 +1,873 @@
+                    break;
+                case Interop.Secur32.ContextAttribute.StreamSizes:
+                    nativeBlockSize = StreamSizes.SizeOf;
+                    break;
+
+                case Interop.Secur32.ContextAttribute.Names:
+                    handleType = typeof(SafeFreeContextBuffer);
+                    break;
+
+                case Interop.Secur32.ContextAttribute.PackageInfo:
+                    handleType = typeof(SafeFreeContextBuffer);
+                    break;
+
+                case Interop.Secur32.ContextAttribute.NegotiationInfo:
+                    handleType = typeof(SafeFreeContextBuffer);
+                    nativeBlockSize = Marshal.SizeOf<NegotiationInfo>();
+                    break;
+
+                case Interop.Secur32.ContextAttribute.ClientSpecifiedSpn:
+                    handleType = typeof(SafeFreeContextBuffer);
+                    break;
+
+                case Interop.Secur32.ContextAttribute.RemoteCertificate:
+                    handleType = typeof(SafeFreeCertContext);
+                    break;
+
+                case Interop.Secur32.ContextAttribute.LocalCertificate:
+                    handleType = typeof(SafeFreeCertContext);
+                    break;
+
+                case Interop.Secur32.ContextAttribute.IssuerListInfoEx:
+                    nativeBlockSize = Marshal.SizeOf<Interop.Secur32.IssuerListInfoEx>();
+                    handleType = typeof(SafeFreeContextBuffer);
+                    break;
+
+                case Interop.Secur32.ContextAttribute.ConnectionInfo:
+                    nativeBlockSize = Marshal.SizeOf<SslConnectionInfo>();
+                    break;
+
+                default:
+                    throw new ArgumentException(SR.Format(SR.net_invalid_enum, "ContextAttribute"), "contextAttribute");
+            }
+
+            SafeHandle SspiHandle = null;
+            object attribute = null;
+
+            try
+            {
+                byte[] nativeBuffer = new byte[nativeBlockSize];
+                errorCode = SecModule.QueryContextAttributes(securityContext, contextAttribute, nativeBuffer, handleType, out SspiHandle);
+                if (errorCode != 0)
+                {
+                    GlobalLog.Leave("Win32:QueryContextAttributes", "ERROR = " + ErrorDescription(errorCode));
+                    return null;
+                }
+
+                switch (contextAttribute)
+                {
+                    case Interop.Secur32.ContextAttribute.Sizes:
+                        attribute = new SecSizes(nativeBuffer);
+                        break;
+
+                    case Interop.Secur32.ContextAttribute.StreamSizes:
+                        attribute = new StreamSizes(nativeBuffer);
+                        break;
+
+                    case Interop.Secur32.ContextAttribute.Names:
+                        attribute = Marshal.PtrToStringUni(SspiHandle.DangerousGetHandle());
```

Comment:
I don't think it's required since `sspiHandle` is defined on the stack. It can't go out of scope as none of these API calls appear to be async.